tap_harvest/__init__.py

```diff
@@ -19,496 +18,35 @@
-
-    # Grab all clients and client contacts. Contacts have client FKs so grab
-    # them last.
-    sync_endpoint("clients")
-    sync_endpoint("contacts", object_to_id=['client'])
-    sync_roles()
-
-    # Sync related project objects
-    sync_endpoint("projects", object_to_id=['client'])
-    sync_endpoint("tasks")
-    sync_endpoint("project_tasks", endpoint='task_assignments', path='task_assignments',
-                  object_to_id=['project', 'task'])
-    sync_endpoint("project_users", endpoint='user_assignments', path='user_assignments',
-                  object_to_id=['project', 'user'])
-
-    # Sync users
-    sync_users()
-
-    if company['expense_feature']:
-        # Sync expenses and their categories
-        sync_endpoint("expense_categories")
-        sync_expenses()
-    else:
-        LOGGER.info("Expense Feature not enabled, skipping.")
-
-    if company['invoice_feature']:
-        # Sync invoices and all related records
-        sync_endpoint("invoice_item_categories")
-        sync_invoices()
-    else:
-        LOGGER.info("Invoice Feature not enabled, skipping.")
-
-    if company['estimate_feature']:
-        # Sync estimates and all related records
-        sync_endpoint("estimate_item_categories")
-        sync_estimates()
-    else:
-        LOGGER.info("Estimate Feature not enabled, skipping.")
-
-    # Sync Time Entries along with their external reference objects
-    sync_time_entries()
-
-    LOGGER.info("Sync complete")
-
 def do_discover():
-    print('{"streams":[]}')
-
-def main_impl():
-    args = utils.parse_args(REQUIRED_CONFIG_KEYS)
-    CONFIG.update(args.config)
-    global AUTH  # pylint: disable=global-statement
-    AUTH = Auth(CONFIG['client_id'], CONFIG['client_secret'], CONFIG['refresh_token'])
-    STATE.update(args.state)
-    # making a copy of STATE for saving child stream bookmark
-    # when data is not available for parent stream
-    TAP_STATE.update(args.state)
-    if args.discover:
-        do_discover()
-    else:
-        do_sync()
-
+    """
+    Call the discovery function.
+    """
+    LOGGER.info('Starting discover')
+    catalog = _discover()
+    json.dump(catalog.to_dict(), sys.stdout, indent=2)
+    LOGGER.info('Finished discover')
+
+@singer.utils.handle_top_exception(LOGGER)
 def main():
-    try:
-        main_impl()
-    except Exception as exc:
-        LOGGER.critical(exc)
-        raise exc
-
-
-if __name__ == "__main__":
+    """
+    Run discover mode or sync mode.
+    """
+    parsed_args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)
+    config = parsed_args.config
+
+    with HarvestClient(config) as client:
+
+        state = {}
+        if parsed_args.state:
+            state = parsed_args.state
+        if parsed_args.discover:
+            do_discover()
+        elif parsed_args.catalog:
+            _sync(client=client,
+                  config=config,
+                  catalog=parsed_args.catalog,
+                  state=state)
```

Comment:
Handle sync mode without catalog, as tap-tester will run sync mode without catalog.
```suggestion
        else:
            catalog = parsed_args.catalog if parsed_args.catalog else do_discover()
            _sync(client=client,
                  config=config,
                  catalog=catalog,
                  state=state)
```

Reply:
Agree @savan-chovatiya . I have updated it to run `discover mode` if `catalog` is not passed in the `sync mode`.